src/parser/paladin/retribution/modules/core/Judgment.js

```diff
@@ -17,38 +21,73 @@ class Judgment extends Analyzer {
   divineStormConsumptions = 0;
   justicarsVengeanceConsumptions = 0;
   judgmentsApplied = 0;
+  judgmentsOverwritten = 0;
+  judgmentTargets = [];
 
   on_byPlayer_applydebuff(event) {
     const spellId = event.ability.guid;
     if (spellId !== SPELLS.JUDGMENT_DEBUFF.id) {
       return;
     }
+    // The event doesn't specify instance on the first target of that type
+    let targetInstance = event.targetInstance;
+    if (targetInstance === undefined) {
+      targetInstance = 1;
+    }
     this.judgmentsApplied += 1;
+    const judgmentTarget = { targetID: event.targetID, targetInstance: targetInstance, timestamp: event.timestamp };
+    this.judgmentTargets.push(judgmentTarget);
   }
 
-  on_byPlayer_damage(event) {
-    const enemy = this.enemies.getEntity(event);
+  on_byPlayer_refreshdebuff(event) {
     const spellId = event.ability.guid;
-
-    if (!enemy) {
+    if (spellId !== SPELLS.JUDGMENT_DEBUFF.id) {
       return;
     }
-    if (!enemy.hasBuff(SPELLS.JUDGMENT_DEBUFF.id, null, 250)) {
+    let targetInstance = event.targetInstance;
+    if (targetInstance === undefined) {
+      targetInstance = 1;
+    }
+    this.judgmentsOverwritten += 1;
+    this.judgmentTargets.forEach((judgmentTarget, index) => {
+      if (judgmentTarget.targetID === event.targetID && judgmentTarget.targetInstance === targetInstance) {
+        judgmentTarget.timestamp = event.timestamp;
+      }
+    });
+  }
+
+  on_byPlayer_damage(event) {
+    const spellId = event.ability.guid;
+    if (!HOLY_POWER_ATTACKS.includes(spellId)) {
       return;
     }
-    switch (spellId) {
-      case SPELLS.TEMPLARS_VERDICT_DAMAGE.id:
-        this.templarsVerdictConsumptions += 1;
-        break;
-      case SPELLS.DIVINE_STORM_DAMAGE.id:
-        this.divineStormConsumptions += 1;
-        break;
-      case SPELLS.JUSTICARS_VENGEANCE_TALENT.id:
-        this.justicarsVengeanceConsumptions += 1;
-        break;
-      default:
-        break;
+    let targetInstance = event.targetInstance;
+    if (targetInstance === undefined) {
+      targetInstance = 1;
     }
+    this.judgmentTargets.forEach((judgmentTarget, index) => {
+      // remove expired judgments
+      const expirationTimestamp = judgmentTarget.timestamp + JUDGMENT_DURATION;
+      if (expirationTimestamp < event.timestamp) {
+        this.judgmentTargets.splice(index, 1);
```

Comment:
Shouldn’t this return then?

Reply:
Just a blank `return;` or the spliced judgmentTarget?
It's important that the array is looked fully through as it is possible to consume multiple judgments with 1 cast.

Reply:
blank return

Reply:
But if you hit someone after it was removed, the spell wouldn't have been buffed